src/GafferSceneUI/SceneGadget.cpp

```diff
@@ -512,13 +502,15 @@ bool SceneGadget::objectAt( const IECore::LineSegment3f &lineInGadgetSpace, Gaff
 		return false;
 	}
 
+	auto viewportGadget = ancestor<ViewportGadget>();
+	const V2f rasterPosition = viewportGadget->gadgetToRasterSpace( lineInGadgetSpace.p1, this );
+	const LineSegment3f offsetLineInGadgetSpace = viewportGadget->rasterToGadgetSpace( rasterPosition + V2f( 0.5f ), this );
```

Comment:
Doing this gadget -> raster + offset -> gadget space conversion adds a noticeable amount of accuracy loss. If I save out the normals for a viewport using this method, and by passing in an already offset `lineInGadgetSpace` (so omitting the extra conversion), the pre-offset one is much better.

I haven't looked at improving precision yet, it could be there's some wins to be had by doing computation in double precision but still passing single precision floats back and forth.

Reply:
I don't fully understand the context here, but why wouldn't we want to pass in an already offset lineInGadgetSpace? I'm assuming this is so that we get a vector that points through the center of the pixel, which seems most correct? Is the reason not to change this just because it would be a compatibility break?

Reply:
Yes, we need a vector that passes through the center of the pixel. We're only given a `GafferUI::ButtonEvent` object in the click handlers, like https://github.com/GafferHQ/gaffer/blob/957ab7fd741232bdcea701ec335d78bc469164c3/src/GafferSceneUI/LightPositionTool.cpp#L770

( `DragDropEvent` derives from `ButtonEvent` ) `ButtonEvent` stores the ray but not the pixel coordinate. So as things are we'll need to do some kind of conversion I guess.

Reply:
I guess it would take a fair bit of an audit to identify everywhere where the ray stored in ButtonEvent is used ... but it seems unlikely that it would ever be desirable to have that ray point to a corner instead of the center?

Reply:
I think Daniel's right here - the ray should already be pointing through the centre of the pixel, long before it ends up in `objectAt()`. Exactly where we put that adjustment is the question. This is the current pipeline as I understand it :

1. We get an event from Qt. From that we get an integer pixel position relative to the widget's origin. Qt does provide a floating point accessor for the position (and in Qt 6 it seems _only_ that), but that is returning integer values too anyway.
2. We convert this into our `ButtonEvent.line`. All our events are floating point, and specify a ray (`.line`) instead of a 2d position, because we want to use them both in Widgets (2d) and Gadgets (3d). At this point our ray points straight into the screen, orthographic-like. But we haven't added a `0.5` offset here.
3. GadgetWidget passes these events to its internal ViewportGadget.
4. ViewportGadget transforms the origin of the line from raster space into 3d space using the camera projection via `rasterToGadgetSpace()`. From there it transforms into the space of whichever gadget it dispatches the event to. That's all done in `eventToGadgetSpace()`.

The places we could put the offset are 2 and 4. I don't think it belongs in `ViewportGadget::rasterToGadgetSpace()` itself because that takes floating point coordinates (if it took integers, that would be a different matter). I think most logically it belongs at 2, the first place we convert from integer coordinates to floating point. That affects a much greater amount of code (all Widgets potentially), so is the riskier choice - I suspect at least one will be thinking in integers. The alternative is 4, applying the offset before calling `rasterToGadgetSpace()`. I don't think it's the _right_ place, but it has the benefit of limiting the knock on effects to only Gadgets.

Since 1.4 is so close, I'm feeling fairly inclined to go with the slightly riskier but more correct option, and just not put it in 1.3.x. But I'm open to other opinions...